### DISCUSSION TITLE
How to do nested comments?

### DISCUSSION THREAD

**Discussion by MuslemRahimi** [pocketbase/pocketbase#2303] [Q&A] [answered]
Hi,
Hope someone can help me out. 

In my case i have a "post" collection where user can write a comment to the related post. This root comment (top level) has a reply feature where user can "Reply" to the comment. I made a "comment" collection (see attachment) and a "replycomment" collection (see attachment). 

Question:

When i now want to get a all comments for a specific post
`const output = await locals.pb.collection("comments").getList(1, 500, {filter: `post="${params.postId}"`, sort: 'created', }) ;`

 how can i get all the related reply comments and also all the reply comments can be also replied to. So that we have an arbitrarily nested reply comments.

![Screenshot from 2023-04-16 18-00-37](https://user-images.githubusercontent.com/22890183/232325580-90715927-a4ac-47a0-943e-74149da3fcac.png)

![Screenshot from 2023-04-16 18-05-53](https://user-images.githubusercontent.com/22890183/232325527-8f5697d9-2fc0-4865-a6f4-9a76e1c0a405.png)
25510-b605b4a5-ce1c-4c0a-bd17-2420732d36d0.png)

**Comment by ganigeorgiev** [comment]
You can use Indirect expand - https://pocketbase.io/docs/expanding-relations/#indirect-expand:

```js
const output = await locals.pb.collection("comments").getList(1, 500, {
    expand: "replycomments(comment)",
    filter: `post="${params.postId}"`,
    sort: 'created'
}) ;
```

But why do you need 2 comments collections (`comments` and `replycomments`) in the first place?
I personally would keep only 1 `comments` collection with optional `replyTo/parent` single `relation` field pointing to the same `comments` collection.

**Reply by MuslemRahimi** [reply]
Thanks for the reply. If i only keep one collections how would the above code change..

**Reply by ganigeorgiev** [reply, marked as answer]
There are several approaches you can take, depending on the desired UX.


<details>
  <summary><strike>One example could be:</strike></summary>
 
_(**edit**: the below example with expand actually will not work in your case because you have the opposite relation child->parent, not parent->child)_

```js
const comments = await locals.pb.collection('comments').getFullList({
    filter: `post="${params.postId}" && parent = ""`, // target the top level comments
    sort: 'created',
    expand: 'parent.parent.parent.parent', // expand up to 4 levels deep
});
```  
</details> 


Alternatively, you can fetch all comments as flatten array and group the comments in a tree structure manually, something like:

```js
// note: I haven't tested the below and you may have to adjust it according to your use case
function listToTree(comments, parentProp = "parent") {
    // create id indexed comments dictionary
    const commentsDict = {};
    for (let comment of comments) {
        commentsDict[comment.id] = {
          comment: comment,
          children: [],
        };
    }

    // build the tree
    const tree = [];
    for (const comment of comments) {
        const parentId = comment[parentProp];
        if (parentId) {
            commentsDict[parentId].children.push(commentsDict[comment.id]);
        } else {
            tree.push(commentsDict[comment.id]);
        }
    }

    return tree;
}

const commentsList = await locals.pb.collection('comments').getFullList({
    filter: `post="${params.postId}"`,
    sort: 'created',
});

const commentsTree = listToTree(commentsList);
```

**Reply by tigawanna** [reply]
A depth field might help too , 0 as default + 1 with every level of nesting

**Reply by shubhanesh0076** [reply]
cool
